code/workspaces/infrastructure-api/src/auth/permissionMiddleware.spec.js

```diff
@@ -0,0 +1,137 @@
+import { permissionTypes } from 'common';
+import { projectPermissionWrapper, systemAdminPermissionWrapper } from './permissionMiddleware';
+
+const { projectPermissions: { PROJECT_KEY_STACKS_LIST }, SYSTEM_INSTANCE_ADMIN } = permissionTypes;
+
+describe('projectPermissionWrapper', () => {
+  describe('when projectKey is provided in request', () => {
+    describe('when user has correct project permissions', () => {
+      const userPermissions = ['projects:testkey:stacks:list'];
+
+      it('extracts and uses the projectKey from the params to pass check', () => {
+        const requestMock = {
+          params: { projectKey: 'testkey' },
+          user: { permissions: userPermissions },
+        };
+        const responseMock = jest.fn();
+        const nextMock = jest.fn();
+
+        projectPermissionWrapper(PROJECT_KEY_STACKS_LIST)(requestMock, responseMock, nextMock);
+
+        expect(nextMock).toHaveBeenCalledTimes(1);
+      });
+
+      it('extracts and uses the projectKey from the body to pass check', () => {
+        const requestMock = {
+          body: { projectKey: 'testkey' },
+          user: { permissions: userPermissions },
+        };
+        const responseMock = jest.fn();
+        const nextMock = jest.fn();
+
+        projectPermissionWrapper(PROJECT_KEY_STACKS_LIST)(requestMock, responseMock, nextMock);
+
+        expect(nextMock).toHaveBeenCalledTimes(1);
+      });
+    });
+
+    describe('when user is system admin without correct project permissions', () => {
+      const userPermissions = [SYSTEM_INSTANCE_ADMIN];
+
+      it('extracts and uses the projectKey from the params to pass check', () => {
+        const requestMock = {
+          params: { projectKey: 'testkey' },
+          user: { permissions: userPermissions },
+        };
+        const responseMock = jest.fn();
+        const nextMock = jest.fn();
+
+        projectPermissionWrapper(PROJECT_KEY_STACKS_LIST)(requestMock, responseMock, nextMock);
+
+        expect(nextMock).toHaveBeenCalledTimes(1);
+      });
+
+      it('extracts and uses the projectKey from the body to pass check', () => {
+        const requestMock = {
+          body: { projectKey: 'testkey' },
+          user: { permissions: userPermissions },
+        };
+        const responseMock = jest.fn();
+        const nextMock = jest.fn();
+
+        projectPermissionWrapper(PROJECT_KEY_STACKS_LIST)(requestMock, responseMock, nextMock);
+
+        expect(nextMock).toHaveBeenCalledTimes(1);
+      });
+    });
+
+    describe('when user does not have correct user permissions or system admin', () => {
+      const userPermissions = ['incorrect:permissions'];
+
+      it('fails with 401 and does not call next', () => {
```

Comment:
The mocking in this test could be replaced by using `node-http-mocks`. Then you can do 
```
const request = httpMocks.createRequest({ params: { projectKey: 'testkey' } });
const response = httpMocks.createResponse();
```

Not sure how you get the `user` object onto the request, you might just be able to add it.